framework/include/utils/SharedPool.h

```diff
@@ -0,0 +1,107 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef SHARED_POOLH
+#define SHARED_POOLH
+
+// System Includes
+#include <stack>
+#include <memory>
+
+namespace MooseUtils
+{
+
+template <class T, typename... Args>
+auto
+reset(int, T & obj, Args... args) -> decltype(obj.reset(args...), void())
```

Comment:
I believe this requires C++14 syntax. Adding the minimum targets to see what happens on our build systems.  Can you comment on why this is needed? If it forces a compiler upgrade, we'll want to know the pros/cons.

Reply:
It's needed to be able to detect whether a `reset()` function exists so we can call it when acquiring an existing object from the pool.

As you noted: this should be legal C++11